cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -177,17 +177,18 @@ struct S3Path {
   std::vector<std::string> key_parts;
 
   static Status FromString(const std::string& s, S3Path* out) {
-    auto first_sep = s.find_first_of(kSep);
+    const auto src = internal::RemoveTrailingSlash(s);
+    auto first_sep = src.find_first_of(kSep);
     if (first_sep == 0) {
       return Status::Invalid("Path cannot start with a separator ('", s, "')");
     }
     if (first_sep == std::string::npos) {
-      *out = {s, s, "", {}};
+      *out = {std::string(src), std::string(src), "", {}};
```

Comment:
(if you prefer; this is not even a nit)
```suggestion
      *out = {src.to_string(), src.to_string(), "", {}};
```